java/src/jni/h5util.c

```diff
@@ -4078,335 +4076,297 @@ translate_rbuf(JNIEnv *env, jobjectArray ret_buf, jlong mem_type_id, H5T_class_t
+
+            typeCount = typeSize / vlSize;
+
+            if (NULL == (objBuf = HDmalloc(typeSize)))
+                H5_OUT_OF_MEMORY_ERROR(ENVONLY, "translate_rbuf: failed to allocate buffer");
+
+            /* Convert each element to a list */
+            for (i = 0; i < (size_t)count; i++) {
+                jboolean     found_jList = JNI_TRUE;
+                jobjectArray jList       = NULL;
+                int          ret_buflen  = -1;
+
+                /* Get the object element */
+                HDmemcpy((char *)objBuf, (char *)raw_buf + i * typeSize, typeSize);
 
-            /* The list we're going to return: */
-            if (NULL == (jList = ENVPTR->GetObjectArrayElement(ENVONLY, (jobjectArray)ret_buf, (jsize)i))) {
-                found_jList = JNI_FALSE;
-                if (NULL == (jList = (jobjectArray)ENVPTR->NewObject(ENVONLY, arrCList, arrListMethod, 0)))
-                    H5_OUT_OF_MEMORY_ERROR(ENVONLY, "translate_rbuf: failed to allocate list read buffer");
+                ret_buflen = ENVPTR->GetArrayLength(ENVONLY, ret_buf);
+                /* The list we're going to return: */
+                if (i < ret_buflen) {
+                    if (NULL ==
+                        (jList = ENVPTR->GetObjectArrayElement(ENVONLY, (jobjectArray)ret_buf, (jsize)i)))
+                        found_jList = JNI_FALSE;
+                }
+                if (NULL == jList) {
+                    if (NULL ==
+                        (jList = (jobjectArray)ENVPTR->NewObject(ENVONLY, arrCList, arrListMethod, 0)))
+                        H5_OUT_OF_MEMORY_ERROR(ENVONLY,
+                                               "translate_rbuf: failed to allocate list read buffer");
+                }
+
+                translate_rbuf(ENVONLY, jList, memb, vlClass, (jsize)typeCount, objBuf);
+                if (!found_jList) {
+                    ENVPTR->CallBooleanMethod(ENVONLY, ret_buf, arrAddMethod, jList);
+                    ENVPTR->DeleteLocalRef(ENVONLY, jList);
+                }
             }
 
-            if ((vlClass == H5T_ARRAY) || (vlClass == H5T_COMPOUND) || (vlClass == H5T_VLEN))
-                translate_rbuf(ENVONLY, jList, memb, vlClass, (jsize)1, objBuf);
-            else {
+            if (objBuf)
+                HDfree(objBuf);
+
+            break;
+        } /* H5T_ARRAY */
+        case H5T_INTEGER: {
+            /* Convert each element to a list */
+            for (i = 0; i < (size_t)count; i++) {
                 jobject jobj = NULL;
-                switch (vlClass) {
-                    /*case H5T_BOOL: {
-                        jboolean boolValue;
-                        for (x = 0; x < vlSize; x++) {
-                            ((char *)&boolValue)[x] = ((char *)vl_elem.p)[j*vlSize+x];
+                switch (typeSize) {
+                    case sizeof(jbyte): {
+                        jbyte byteValue;
+                        for (x = 0; x < typeSize; x++) {
```

Comment:
These loops probably should be changed to memcpy, maybe?

Reply:
Yes, I think it would be more appropriate to just do a memcpy of size `sizeof(jbyte)`. It's still dangerous if typeSize isn't correct, but the loops below are somewhat misleading and may not necessarily be correct.

Reply:
Done for next PR.